particle-iot/device-os
# Allowing the ALT server public key to be set in DCT programatically

### Problem

The documentation for `System Config [ set ]` suggests that the public server key for a device (including the host and key) can be updated from user applications as follows:

```
System.set(SYSTEM_CONFIG_SERVER_KEY, newKeyBytes, newKeyString.length());
```

This works as expected on Photon but was never updated for UDP devices.

### Solution

This PR updates the `stm32f2xx` and `nRF52840` HAL layers to update the server public key stored in DCT at `DCT_ALT_SERVER_PUBLIC_KEY_OFFSET` on UDP devices instead of updating the seemingly unused `DCT_SERVER_PUBLIC_KEY_OFFSET` location.

### Steps to Test

1. Generate a hex string for the production key file: `xxd -p -c 768 production.pub.der`
2. Insert that hex string into the test app, compile, and flash it to the test device.
3. Point test device at non-production environment using `particle server key <non-production-keyfile>`
4. Once the device has connected to the non-production environment, call the `change-server` function.
5. Device should update the key locally, reset itself, and connect to the production environment.

### Example App

```c
bool keyLoadRequired = false;
bool resetRequired = false;

int setServerKey(String unused) {
  keyLoadRequired = true;
  return 1;
}

void loadNewKey() {
  String newKeyString = "<key hex string>"; // Use: xxd -p -c 768 production.pub.der
  char newKeyBytes[768] = {0};
  for (unsigned int i = 0, j = 0; i < newKeyString.length(); i += 2, j += 1) {
    String byteString = newKeyString.substring(i, i + 2);
    newKeyBytes[j] = (char) strtol(byteString.c_str(), NULL, 16);
  }
  System.set(SYSTEM_CONFIG_SERVER_KEY, newKeyBytes, newKeyString.length());

  Particle.publish("dut-moving", "production", 60);
  resetRequired = true;
}

void setup() {
  Particle.function("change-server", setServerKey);
  delay(500);
  Particle.publish("dut-boot", "complete", 60);
}

void loop() {
  if (keyLoadRequired) {
    keyLoadRequired = false;
    loadNewKey();
  }

  if (resetRequired) {
    resetRequired = false;
    System.reset();
  }
}
```

### References

* [ch61883]
